src/vsr/client.zig

```diff
@@ -26,10 +29,97 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
                 operation: StateMachine.Operation,
                 results: []const u8,
             ) void;
-            user_data: u128,
-            // Null iff operation=register.
-            callback: ?Callback,
+
+            const Info = struct {
+                user_data: u128,
+                callback: ?Callback, // Null iff operation=register.
+            };
+
+            const Demux = struct {
+                next: ?*Demux = null,
+                info: Info,
+                event_count: u16,
+                event_offset: u16,
+            };
+
+            info: Info,
             message: *Message,
+            demux_queue: FIFO(Demux) = .{ .name = null },
+        };
+
+        pub const Batch = struct {
+            request: *Request,
+            demux: ?*Request.Demux, // Null iff batch_get() newly created `request`
+
+            pub fn slice(batch: Batch) []u8 {
+                const message = batch.request.message;
+                assert(message.header.command == .request);
+
+                const body_offset = if (batch.demux) |demux| blk: {
+                    const operation = message.header.operation.cast(StateMachine);
+                    switch (operation) {
+                        inline else => |op| {
+                            const event_size: u32 = @sizeOf(StateMachine.Event(op));
+                            break :blk event_size * demux.event_offset;
+                        },
+                    }
+                } else blk: {
+                    assert(batch.request.demux_queue.empty());
```

Comment:
We can also assert the inverse `assert(!batch.request.demux_queue.empty())` when a `demux` is present.

Maybe we could assert if the `demux` is contained in the `request.demux_queue` (enclosed by `if (constants.verify)`)